### PR TITLE
Background job time windows

### LINKED ISSUE #30273: Timed-based background jobs
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are interested into the same feature.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

**Is your feature request related to a problem? Please describe.**
Some background jobs are long-runners and create heavy load, this can also happen during office hours impacting the system's performance for users during such times.

**Describe the solution you'd like**
Provide a way to define a time when a background job should run.
e.g. to ensure heavy load jobs don't run during office hours but rather at night.

**Describe alternatives you've considered**
none

**Additional context**
-

### PR DESCRIPTION
**Pull request by nickvergessen**
- [ ] Should be used by background jobs which are rather heavy but are not time critical:
   - [x] Background jobs in server code
   - [ ] Activity expiration
   - [ ] Suspicious login
   - [ ] Mail ?
   - [ ] Talk ?
- [x] Write documentation for app developers
- [x] Write documentation for sysadmins

### CODE REVIEW

File: config/config.sample.php [hunk 1]
```diff
@@ -1214,6 +1214,20 @@
...
 
+/**
+ * UTC Hour for maintenance windows
+ *
+ * Some background jobs only run once a day. When an hour is defined for this config,
+ * the background jobs which advertise themselves as not time sensitive will be
+ * delayed during the "working" hours and only run in the 4 hours after the given time.
+ * This is e.g. used for activity expiration, suspicious login training and update checks.
+ *
+ * A value of 1 e.g. will only run this background jobs between 01:00am UTC and 05:00am UTC.
```
Comment (nickvergessen) [inline config/config.sample.php:1225]:
```suggestion
 * A value of 1 e.g. will only run these background jobs between 01:00am UTC and 05:00am UTC.
```

**Comment by PVince81** [comment]
@nickvergessen I've assigned you since you worked on this, let me know if you're not able to finish it and I can find someone else, let me know

it looks like the base work is already done, the decision/discussion for switching individual background jobs to the new mode could be done separately if needed

**Comment by nickvergessen** [comment]
> I've assigned you since you worked on this, let me know if you're not able to finish it and I can find someone else, let me know
> 
> it looks like the base work is already done, the decision/discussion for switching individual background jobs to the new mode could be done separately if needed

Yes, it's basically done, just missing the developer documentation and a note in the upgrade guideline for admins

**Comment by nickvergessen** [comment]
Rebased and squashed

**Review by ChristophWurst — APPROVED** [review]
Very clean :shipit:

**Review by miaulalala — APPROVED** [review]
:rocket:

**Comment by AndyScherzinger** [comment]
would also need backporting to v22 and up - cc @PVince81

**Comment by PVince81** [comment]
@nickvergessen can you update the docs for devs ?

**Comment by nickvergessen** [comment]
> can you update the docs for devs ?

https://github.com/nextcloud/documentation/pull/8087

**Pull request merged**